Clustering/Hdbscan/src/main/java/org/tribuo/clustering/hdbscan/HdbscanTrainer.java

```diff
@@ -0,0 +1,828 @@
+        }
+        return distance;
+    }
+
+    @Override
+    public String toString() {
+        return "HdbscanTrainer(minClusterSize=" + minClusterSize + ",distanceType=" + distanceType + ",k=" + k + ",numThreads=" + numThreads + ")";
+    }
+
+    @Override
+    public TrainerProvenance getProvenance() {
+        return new TrainerProvenanceImpl(this);
+    }
+
+    /**
+     * A cluster exemplar, with attributes for the point's label, outlier score and its features.
+     */
+    static class ClusterExemplar {
+        private final Integer label;
+        private final Double outlierScore;
+        private final SGDVector features;
+
+        ClusterExemplar(Integer label, Double outlierScore, SGDVector features) {
+            this.label = label;
+            this.outlierScore = outlierScore;
+            this.features = features;
+        }
+
+        Integer getLabel() {
+            return label;
+        }
+
+        Double getOutlierScore() {
+            return outlierScore;
+        }
+
+        SGDVector getFeatures() {
+            return features;
+        }
+    }
+
+    /**
+     * A Runnable implementation of the core distance calculation for parallelization.
+     * To be used with an {@link ExecutorService}
+     */
+    private static class CoreDistanceRunnable implements Runnable {
+
+        final private SGDVector[] data;
+        final private int numNeighbors;
+        final private Distance distanceType;
+        final private int point;
+        DenseVector coreDistances;
+
+        CoreDistanceRunnable(SGDVector[] data, int numNeighbors, Distance distanceType, int point, DenseVector coreDistances) {
+            this.data = data;
+            this.numNeighbors = numNeighbors;
+            this.distanceType = distanceType;
+            this.point = point;
+            this.coreDistances = coreDistances;
+        }
+
+        @Override
+        public void run() {
+            // This logic is duplicated in the calculateCoreDistances method in the outer class above
+            double[] kNNDistances = new double[numNeighbors];
+            for (int i = 0; i < numNeighbors; i++) {
+                kNNDistances[i] = Double.MAX_VALUE;
+            }
+
+            for (int neighbor = 0; neighbor < data.length; neighbor++) {
+                if (point == neighbor)
+                    continue;
+                double distance = getDistance(data[point], data[neighbor], distanceType);
+
+                // Check at which position in the nearest distances the current distance would fit
+                int neighborIndex = numNeighbors;
+                while (neighborIndex >= 1 && distance < kNNDistances[neighborIndex-1]) {
+                    neighborIndex--;
+                }
+
+                // Shift elements in the array to make room for the current distance
+                if (neighborIndex < numNeighbors) {
+                    for (int shiftIndex = numNeighbors-1; shiftIndex > neighborIndex; shiftIndex--) {
```

Comment:
Similar comments to the version above about use of `Arrays.binarySearch` and `System.arraycopy`.